src/client.rs

```diff
@@ -160,38 +251,272 @@ where
+    /// # use futures_executor::block_on;
+    /// # use std::iter::once;
+    /// # use tuf::{
+    /// #     Error,
+    /// #     interchange::Json,
+    /// #     client::{Client, Config},
+    /// #     crypto::{KeyType, PrivateKey, SignatureScheme},
+    /// #     metadata::{MetadataPath, MetadataVersion, Role, RootMetadataBuilder},
+    /// #     repository::{EphemeralRepository, Repository},
+    /// # };
+    /// # fn main() -> Result<(), Error> {
+    /// # block_on(async {
+    /// # let private_key = PrivateKey::from_pkcs8(
+    /// #     &PrivateKey::new(KeyType::Ed25519)?,
+    /// #     SignatureScheme::Ed25519,
+    /// # )?;
+    /// # let public_key = private_key.public().clone();
+    /// let local = EphemeralRepository::new();
+    /// let remote = EphemeralRepository::new();
+    ///
+    /// let root_version = 1;
+    /// let root_threshold = 1;
+    /// let root = RootMetadataBuilder::new()
+    ///     .version(root_version)
+    ///     .expires(Utc.ymd(2038, 1, 1).and_hms(0, 0, 0))
+    ///     .root_key(public_key.clone())
+    ///     .root_threshold(root_threshold)
+    ///     .snapshot_key(public_key.clone())
+    ///     .targets_key(public_key.clone())
+    ///     .timestamp_key(public_key.clone())
+    ///     .signed::<Json>(&private_key)?;
+    ///
+    /// let root_path = MetadataPath::from_role(&Role::Root);
+    /// let root_version = MetadataVersion::Number(root_version);
+    ///
+    /// remote.store_metadata(&root_path, &root_version, &root).await?;
+    ///
+    /// let client = Client::from_pinned_root_keys(
+    ///     Config::default(),
+    ///     &root_version,
+    ///     root_threshold,
+    ///     once(&public_key),
+    ///     local,
+    ///     remote,
+    /// ).await?;
+    /// # Ok(())
+    /// # })
+    /// # }
+    /// ```
+    pub async fn from_pinned_root_keys<'a, I>(
+        config: Config<T>,
+        root_version: &MetadataVersion,
+        root_threshold: u32,
+        trusted_root_keys: I,
+        local: L,
+        remote: R,
+    ) -> Result<Self>
+    where
+        I: IntoIterator<Item = &'a PublicKey>,
+    {
+        let root_path = MetadataPath::from_role(&Role::Root);
+
+        let (fetched, root) = fetch_metadata_from_local_then_remote(
+            &root_path,
+            root_version,
+            config.max_root_length,
+            &local,
+            &remote,
+        )
+        .await?;
+
+        let tuf = Tuf::from_root_keys_pinned(root.clone(), root_threshold, trusted_root_keys)?;
+
+        // Only store the metadata after we have validated it.
+        if fetched {
+            let root_version = MetadataVersion::Number(root.version());
+            local
+                .store_metadata(&root_path, &root_version, &root)
+                .await?;
+
+            // Also store this root metadata as the latest version.
+            local
+                .store_metadata(&root_path, &MetadataVersion::None, &root)
+                .await?;
+        }
```

Comment:
I think this should be conditional on whether you're increasing the version somehow. Can you add tests for this case to each of these constructors to show that these don't roll back root?